File: java/224.java
import java.util.Stack;

class Solution {
    public int calculate(String s) {
        Stack<Integer> num_stack = new Stack<>();
        Stack<Character> ops_stack = new Stack<>();
        int length = s.length();
        int i = 0;
        while (i < length) {
            if (s.charAt(i) == '(') {
                ops_stack.push('(');
            } else if (s.charAt(i) == ')') {
                char op;
                while ((op = ops_stack.pop()) != '(') {
                    if (op == '+') {
                        int result = num_stack.pop() + num_stack.pop();
                        num_stack.push(result);
                    } else {
                        int top1 = num_stack.pop();
                        int top2 = num_stack.pop();
                        int result = top2 - top1;
                        num_stack.push(result);
                    }
                }
            } else if (s.charAt(i) == '+') {
                if (!ops_stack.isEmpty()) {
                    char op = ops_stack.peek();
                    if (op == '+') {
                        int result = num_stack.pop() + num_stack.pop();
                        num_stack.push(result);
                    } else if (op == '-') {
                        int top1 = num_stack.pop();
                        int top2 = num_stack.pop();
                        int result = top2 - top1;
                        num_stack.push(result);
                        ops_stack.pop();
                        ops_stack.push('+');
                    } else {
                        ops_stack.push('+');
                    }
                } else {
                    ops_stack.push('+');
                }
            } else if (s.charAt(i) == '-') {
                if (!ops_stack.isEmpty()) {
                    char op = ops_stack.peek();
                    if (op == '+') {
                        int result = num_stack.pop() + num_stack.pop();
                        ops_stack.pop();
                        num_stack.push(result);
                        ops_stack.push('-');
                    } else if (op == '-') {
                        int top1 = num_stack.pop();
                        int top2 = num_stack.pop();
                        int result = top2 - top1;
                        num_stack.push(result);
                    } else {
                        ops_stack.push('-');
                    }
                } else {
                    ops_stack.push('-');
                }
            } else if (s.charAt(i) == ' ') {
                i++;
                continue;
            } else {
                int num = s.charAt(i) - '0';
                i += 1;
                while (i < length) {
                    if (s.charAt(i) >= '0' && s.charAt(i) <= '9') {
                        int n = s.charAt(i) - '0';
                        num = num * 10 + n;
                        i++;
                    } else {
                        break;
                    }
                }
                num_stack.push(num);
                i--;
            }
            i++;
        }
        while (!ops_stack.isEmpty()) {
            char c = ops_stack.pop();
            int num1 = num_stack.pop();
            int num2 = num_stack.pop();
            if (c == '+') {
                int result = num1 + num2;
                num_stack.push(result);
            } else {
                int result = num2 - num1;
                num_stack.push(result);
            }
        }
        return num_stack.pop();
    }

    public static void main(String[] args) {
        Solution s = new Solution();
        String str = "(1-(3-4))";
        int result = s.calculate(str);
        System.out.println(result);
    }
}

File: python/224.py
class Solution:
    def calculate(self, s):
        num_stack = []
        ops_stack = []
        length = len(s)
        i = 0
        while i < length:
            c = s[i]
            if c == '(':
                ops_stack.append(c)
            elif c == ')':
                op = ops_stack.pop()
                while op != '(':
                    num1 = num_stack.pop()
                    num2 = num_stack.pop()
                    result = num2 + num1 if op == '+' else num2 - num1
                    num_stack.append(result)
                    op = ops_stack.pop()
            elif c == '+':
                if not ops_stack:
                    ops_stack.append(c)
                elif ops_stack[-1] == '+':
                    num1 = num_stack.pop()
                    num2 = num_stack.pop()
                    result = num2 + num1
                    num_stack.append(result)
                elif ops_stack[-1] == '-':
                    ops_stack.pop()
                    num1 = num_stack.pop()
                    num2 = num_stack.pop()
                    result = num2 - num1
                    num_stack.append(result)
                    ops_stack.append(c)
                elif ops_stack[-1] == '(':
                    ops_stack.append(c)
            elif c == '-':
                if not ops_stack:
                    ops_stack.append(c)
                elif ops_stack[-1] == '+':
                    ops_stack.pop()
                    num1 = num_stack.pop()
                    num2 = num_stack.pop()
                    result = num2 + num1
                    num_stack.append(result)
                    ops_stack.append(c)
                elif ops_stack[-1] == '-':
                    num1 = num_stack.pop()
                    num2 = num_stack.pop()
                    result = num2 - num1
                    num_stack.append(result)
                elif ops_stack[-1] == '(':
                    ops_stack.append(c)
            elif c == ' ':
                i += 1
                continue
            else:
                num = int(c)
                i += 1
                while i < length:
                    c = s[i]
                    if c >= '0' and c <= '9':
                        num = num * 10 + int(c)
                        i += 1
                    else:
                        break
                i -= 1
                num_stack.append(num)
            i += 1
        while ops_stack:
            num1 = num_stack.pop()
            num2 = num_stack.pop()
            op = ops_stack.pop()
            if op == '+':
                result = num2 + num1
            else:
                result = num2 - num1
            num_stack.append(result)
        return num_stack.pop()
